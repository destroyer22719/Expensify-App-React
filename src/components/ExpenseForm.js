import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

const now = moment();

class ExpenseForm extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount/100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocus: false,
      errorState: '',
    }
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({description}));
  }
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  }
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount}))
    }
  }
  onDateChange = (date) => {
    if (date) this.setState(() => ({createdAt: date}))
  }
  onFocusChange = ({focused}) => {
    this.setState(() => ({
     calendarFocus: focused
    }))
  }
  onSubmit = e => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        errorState: 'Cannot can have description or amount empty'
      }))
    } else {
      this.setState(() => ({errorState: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: (+this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      })
    }
  }
  render() {
    return (
      <div>
        {<div>{this.state.errorState}</div>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="Description" 
            autoFocus 
            value={this.state.description}
            onChange={this.onDescriptionChange}
            />
          <input 
            type="number" 
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}  
          />
          <SingleDatePicker 
            id="SingleDatePicker"
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocus}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />
          <textarea 
            placeholder="Add a note for your expense (option)" 
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}

export default ExpenseForm;