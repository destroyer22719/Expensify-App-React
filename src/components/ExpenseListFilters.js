import React from 'react';
import { connect } from 'react-redux';
import { setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'

class ExpenseListFilters extends React.Component {

  state = {
    calendarFocus: null,
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  }
  onFocusChange = (calendarFocus) => {
    this.setState(() => {
      this.setState(() => ({
        calendarFocus
      }))
    })
  }
  render() {
    console.log(this.props)
    return (
      <div>
      <input 
        type="text" 
        value={this.props.filters.text} 
        onChange={e => {
          this.props.dispatch(setTextFilter(e.target.value))
      }}/>
      <select
        onChange={(e) => {
          if (e.target.value === "amount") this.props.dispatch(sortByAmount());
          else if (e.target.value === "date") this.props.dispatch(sortByDate());
        }}
      >
        <option value="date">Date</option>
        <option value="amount" >Amount</option>
      </select>
      <DateRangePicker 
        startDate={this.props.filters.startDate}
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocus}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}
        showClearDates={true}
      />
    </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filter
  }
}

export default connect(mapStateToProps)(ExpenseListFilters);