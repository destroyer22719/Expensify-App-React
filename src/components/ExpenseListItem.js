import React from "react";
import { connect } from "react-redux";
import { Link, NavLink } from "react-router-dom";
import { startRemoveExpense } from "../actions/expenses";
import moment from "moment";
import numeral from "numeral";

export const ExpenseListItem = ({
    startRemoveExpense,
    description,
    id,
    amount,
    createdAt,
}) => (
    <Link className="list-item" to={`/edit/${id}`}>
        <div>
            <h3 className="list-item__title">{description}</h3>
            <span className="list-item__sub-title">
                {moment(createdAt).format("MMMM Do, YYYY")}
            </span>
        </div>
        <h3 className="list-item__data">
            {numeral(amount / 100).format("$0,0.00")}
        </h3>
    </Link>
);

const matchDispatchToProps = (dispatch) => ({
    startRemoveExpense: (id) => dispatch(startRemoveExpense({ id })),
});

export default connect(null, matchDispatchToProps)(ExpenseListItem);
