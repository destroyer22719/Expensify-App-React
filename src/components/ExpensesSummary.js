import React from 'react';

export const ExpensesSummary = props => (
  <div>
    <h2>Viewing {props.expensesCount} totalling $ {props.expensesTotal}</h2>
  </div>
)