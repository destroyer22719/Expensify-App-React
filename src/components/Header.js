import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth';

const Header = ({startLogout}) => (
  <header>
    <h1>Expensify App</h1>
    <ul>
      <li><NavLink to="/dashboard" activeClassName="active">Dashboard</NavLink></li>
      <li><NavLink to="/create" activeClassName="active">Create</NavLink></li>
      <li><NavLink to="/help" activeClassName="active" >Help</NavLink></li>
    </ul>
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);