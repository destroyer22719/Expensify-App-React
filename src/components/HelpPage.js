import React from 'react';

export const HelpPage = () => (
  <div>This is the Help Page</div>
)

export default HelpPage