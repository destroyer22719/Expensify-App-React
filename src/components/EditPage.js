import React from "react";
import { connect } from "react-redux";
import { startRemoveExpense, startEditExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class EditPage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button
                        className="button button--secondary"
                        onClick={this.onRemove}
                    >
                        Remove Expense
                    </button>
                </div>
            </div>
        );
    }
}

// const EditPage = (props) => {
//   console.log(props.expense)
//   return (
//     <div>
//     {!props.expense && <p>Not Found</p>}
//       {
//         props.expense && <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(startEditExpense(props.expense.id, {...expense}))
//           props.history.push('/')
//         }}
//     />}
//     </div>
// )}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expenses) => {
            return expenses.id === props.match.params.id;
        }),
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (expenseId, expense) =>
        dispatch(startEditExpense(expenseId, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);
