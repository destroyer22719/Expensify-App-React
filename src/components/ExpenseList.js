import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import { getExpensesTotal } from '../selectors/expenses-total';
import ExpenseListItem from './ExpenseListItem';
import { ExpensesSummary } from './ExpensesSummary';

export const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    {
      !props.expenses.length ? (
        <p>No Expenses</p>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense}/>
        })
      )
    }
    <ExpensesSummary expenseCount={props.expenses.length} expensesTotal={getExpensesTotal(props.expenses)/100}/>
  </div>
);

const mapStateToProps = (state) => {
  console.log(state);
  return {
    expenses: getVisibleExpenses(state.expenses, state.filter)
  }
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList;