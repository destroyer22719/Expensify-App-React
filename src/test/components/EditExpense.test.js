import React from 'react';
import {shallow} from 'enzyme';
import '../setupTest';
import { expensesData } from '../fixtures/expenses.data';
import { EditPage } from '../../components/EditPage';

let editExpense, history, removeExpense, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn()};
  wrapper = shallow(
    <EditPage 
      editExpense={editExpense} 
      history={history}
      removeExpense={removeExpense}
      expense={expensesData[1]}
    />)
});

test('should render EditPage', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should handle edit expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expensesData[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expensesData[1].id, expensesData[1])
})

test('should handle remove expense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({id: expensesData[1].id})
})

test('', () => {
  
})