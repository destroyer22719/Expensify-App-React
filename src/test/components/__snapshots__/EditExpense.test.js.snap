// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPage 1`] = `
<div>
  <div>
    <ExpenseForm
      expense={
        Object {
          "amount": 2000,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
